components/datetime/src/format/datetime.rs

```diff
@@ -87,67 +92,82 @@ where
     }
 }
 
-// Temporary formatting number with length.
-fn format_number<W>(result: &mut W, num: isize, length: FieldLength) -> Result<(), core::fmt::Error>
+// Apply length to input number and write to result using fixed_decimal_format.
+fn format_number<W>(
+    result: &mut W,
+    fixed_decimal_format: &FixedDecimalFormat,
+    mut num: FixedDecimal,
+    length: FieldLength,
+) -> fmt::Result
 where
     W: fmt::Write + ?Sized,
 {
     match length {
-        FieldLength::One => write!(result, "{}", num),
+        FieldLength::One => {}
         FieldLength::TwoDigit => {
-            if num < 100 {
-                write!(result, "{:0>width$}", num, width = 2)
-            } else {
-                let buffer = num.to_string();
-                let len = buffer.len();
-                // Safe because we've handled the case where len < 2 above
-                #[allow(clippy::indexing_slicing)]
-                result.write_str(&buffer[len - 2..])
-            }
+            num.pad_left(2);
+            num.truncate_left(1);
+        }
+        FieldLength::Abbreviated => {
+            num.pad_left(3);
+        }
+        FieldLength::Wide => {
+            num.pad_left(4);
+        }
+        FieldLength::Narrow => {
+            num.pad_left(5);
+        }
+        FieldLength::Six => {
+            num.pad_left(6);
         }
-        FieldLength::Abbreviated => write!(result, "{:0>width$}", num, width = 3),
-        FieldLength::Wide => write!(result, "{:0>width$}", num, width = 4),
-        FieldLength::Narrow => write!(result, "{:0>width$}", num, width = 5),
-        FieldLength::Six => write!(result, "{:0>width$}", num, width = 6),
         FieldLength::Fixed(p) => {
-            let buffer = num.to_string();
-            let len = buffer.len();
-            if len < p.into() {
-                write!(result, "{:0<width$}", num, width = p as usize)
-            } else {
-                // Safe because we've handled the case where len < p above
-                #[allow(clippy::indexing_slicing)]
-                result.write_str(&buffer[0..p as usize])
-            }
+            num.pad_left(p as u16);
+            num.truncate_left((p - 1) as i16);
```

Comment:
> thought

I assume because `FieldLength::One` is covered that `p - 1` cannot underflow, but I just want to be sure.

Reply:
This should be `p as i16 - 1`, thanks for noticing this!